src/python/pants/jvm/package/war.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+import textwrap
+from dataclasses import dataclass
+from pathlib import PurePath
+
+from pants.base.glob_match_error_behavior import GlobMatchErrorBehavior
+from pants.build_graph.address import Address
+from pants.core.goals.package import (
+    BuiltPackage,
+    BuiltPackageArtifact,
+    OutputPathField,
+    PackageFieldSet,
+)
+from pants.core.target_types import ResourceSourceField, FileSourceField
+from pants.core.util_rules.archive import ZipBinary, UnzipBinary
+from pants.core.util_rules.source_files import SourceFilesRequest, SourceFiles
+from pants.core.util_rules.stripped_source_files import StrippedSourceFiles
+from pants.engine.addresses import Addresses, UnparsedAddressInputs
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, FileContent, MergeDigests, DigestSubset, PathGlobs, \
+    GlobExpansionConjunction, DigestEntries, FileEntry, DigestContents
+from pants.engine.internals.native_engine import EMPTY_DIGEST
+from pants.engine.internals.selectors import MultiGet
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import DependenciesRequest, Targets, SourcesField, WrappedTarget, HydratedSources, \
+    HydrateSourcesRequest
+from pants.engine.unions import UnionRule
+from pants.jvm.classpath import Classpath
+from pants.jvm.target_types import JvmWarDescriptorAddressField, JvmWarDependenciesField, JvmWarContentField
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class PackageWarFileFieldSet(PackageFieldSet):
+    required_fields = (
+        JvmWarDependenciesField,
+        JvmWarDescriptorAddressField,
+    )
+
+    output_path: OutputPathField
+    dependencies: JvmWarDependenciesField
+    descriptor: JvmWarDescriptorAddressField  # TODO: make required
+    content: JvmWarContentField
+
+
+@dataclass(frozen=True)
+class RenderDeploymentDescriptorRequest:
+    descriptor: JvmWarDescriptorAddressField
+    owning_address: Address
+
+
+@dataclass(frozen=True)
+class RenderedDeploymentDescriptor:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class RenderContentRequest:
```

Comment:
given Pants house style is to use non-namespaced imports, it'll probably make it easier to read in future if the `Render` classnames all have `War` in them somewhere.

Reply:
Done.